docs(providers): update providers documentation

0ubbe opened the pull request:
<!--
Thanks for your interest in the project. Bugs filed, and PRs submitted are appreciated!

Please make sure that you are familiar with and follow the Code of Conduct for
this project (found in the CODE_OF_CONDUCT.md file).

Also, please make sure you're familiar with and follow the instructions in the
contributing guidelines (found in the CONTRIBUTING.md file).

If you're new to contributing to open source projects, you might find this free
video course helpful: https://kcd.im/pull-request

Please fill out the information below to expedite the review and (hopefully)
merge of your pull request!
-->

## Reasoning 💡

[We realised](https://github.com/nextauthjs/next-auth/pull/1876/files/6df72b692e3e6427df3967be949b5265b8c76d4f#diff-7a1af8aab4c41d70f23eaf05d5d4fcf5f969a20f515122828f970ab3b008bd69) that the documentation for the different available provider options was out of date.

I took the opportunity to re-work a bit the **Providers** page in our docs to delineate better the 3 ways to sign-in

- `OAuth`
- `email`
- `credentials`

and the steps and options for each. When listing the options for each type of provider, it took [our types](https://github.com/nextauthjs/next-auth/blob/main/types/providers.d.ts) as the source of truth, as agreed with @balazsorban44. I also added emojis 🙈

<img width="600" alt="Screenshot 2021-04-30 at 18 07 23" src="https://user-images.githubusercontent.com/5938217/116723188-86d77380-a9df-11eb-9360-327ecd3be8c5.png">


## Checklist 🧢

- [x] Documentation
- [ ] ~Tests~
- [x] Ready to be merged

## Affected issues 🎟

#1792

0ubbe committed:
docs(providers): update providers documentation

- delineate clearly the 3 provider types (oauth, email, credentials)
- make each section structure consistent
- update the option list for every provider type
- use emojis

balazsorban44 reviewed (changes requested):
https://next-auth-git-docs-1792-providers-params-table-nextauthjs.vercel.app/configuration/providers#-adding-a-new-provider

could you mention adding the type in types/providers to be able to see it in TS as step 3? (or we could figure out how to generate that type out of the src/providers folder, as we generate the index.js file as well)

0ubbe committed:
Merge branch 'main' into docs/1792-providers-params-table

0ubbe committed:
docs(providers): instructions on new provider types

0ubbe committed:
docs(providers): remove emojis

To stay consistent with the rest of our documentation, for now we should not emojis on the sections of our documentation pages.

0ubbe committed:
Merge branch 'main' into docs/1792-providers-params-table

balazsorban44 commented on www/docs/configuration/providers.md:
```diff
@@ -3,13 +3,22 @@ id: providers
+### Available providers
 
 <ul>
 {Object.entries(require("../../providers.json"))
```
This list at
https://next-auth-git-docs-1792-providers-params-table-nextauthjs.vercel.app/configuration/providers#available-providers

is getting pretty long. Maybe instead of `ul>li`, we should just have a comma-separated list of the providers similar to how it is done here: https://next-auth-git-docs-1792-providers-params-table-nextauthjs.vercel.app/faq#what-authentication-services-does-nextauthjs-support

lluia replied:
Nice 💚 , this is how it's gonna look:

<img width="600" alt="Screenshot 2021-05-02 at 21 14 31" src="https://user-images.githubusercontent.com/5938217/116824606-64825900-ab8b-11eb-90da-0de22e6a6870.png">

balazsorban44 commented on www/docs/configuration/providers.md:
```diff
@@ -20,42 +29,73 @@ NextAuth.js is designed to work with any OAuth service, it supports OAuth 1.0, 1
+|        type         |              Type of provider, in this case `oauth`              |              `oauth`              |   Yes    |
+|       version       |            OAuth version (e.g. '1.0', '1.0a', '2.0')             |             `string`              |   Yes    |
+|        scope        |          OAuth access scopes (expects array or string)           |      `string` or `string[]`       |   Yes    |
+|       params        |             Additional authorization URL parameters              |     `{ grant_type: string }`      |   Yes    |
```
I think the description is wrong on this one (that said, the name is also too generic). I believe this is used to send extra params to the `/token` endpoint when getting an access_token from the OAuth provider. The type is not limited to `grant_type` either.

We do some magic :rainbow: with the params (open the link and scroll, there is more than in this snippet.): https://github.com/nextauthjs/next-auth/blob/dfe4620056b6ba3f72d1dd5ab166efb7f29db5c7/src/server/lib/oauth/client.js#L102-L115

We do some provider-specific things here that shouldn't be in the core, and I hope that #1846 will clear this up because right now it is quite confusing and not flexible enough.

balazsorban44 commented on www/docs/configuration/providers.md:
````diff
@@ -89,7 +129,8 @@ As an example of what this looks like, this is the provider object returned for
 ```
-You can replace all the options in this JSON object with the ones from your custom provider - be sure to give it a unique ID and specify the correct OAuth version - and add it to the providers option:
+
+You can replace all the options in this JSON object with the ones from your custom provider - be sure to give it a unique ID and specify the correct OAuth version - and add it to the providers option when initializing the library:
````
```suggestion
Replace all the options in this JSON object with the ones from your custom provider - be sure to give it a unique ID and specify the correct OAuth version - and add it to the providers option when initializing the library:
```

balazsorban44 commented on www/docs/configuration/providers.md:
```diff
@@ -20,42 +29,73 @@ NextAuth.js is designed to work with any OAuth service, it supports OAuth 1.0, 1
+|       region        |                    Only when using BattleNet                     |             `string`              |    No    |
+|       domain        |                Only when using certain Providers                 |             `string`              |    No    |
+|      tenantId       |     Only when using Azure, Active Directory, B2C, FusionAuth     |             `string`              |    No    |
+
```
```suggestion
:::tip
Even if you are using a built-in provider, you can override any of these options to tweak the default configuration.

```js title=[...nextauth].js
import Providers from "next-auth/providers"
...
Providers.Auth0({
  clientId: process.env.CLIENT_ID,
  clientSecret: process.env.CLIENT_SECRET,
  domain: process.env.DOMAIN,
  scope: "openid your_custom_scope", // We do provide a default, but this will override it if defined
  profile(profile) {
    return {} // Return the profile in a shape that is different from the built-in one.
  }
})
...
:::
```

0ubbe committed:
docs(providers): reword sentence

Co-authored-by: Balázs Orbán <info@balazsorban.com>

0ubbe committed:
docs(providers): add tip on overriding options

0ubbe committed:
docs(providers): clarify `params` option usage

0ubbe committed:
docs(providers): make names list inline

0ubbe committed:
Merge branch 'main' into docs/1792-providers-params-table

lluia commented:
@balazsorban44 addressed all your suggestions 🙇🏽‍♂️

balazsorban44 reviewed (approved):
Awesome!
